internal/controller/nicdevice_controller.go

```diff
@@ -0,0 +1,550 @@
+/*
+2024 NVIDIA CORPORATION & AFFILIATES
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	k8sTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1alpha1 "github.com/Mellanox/nic-configuration-operator/api/v1alpha1"
+	"github.com/Mellanox/nic-configuration-operator/pkg/consts"
+	"github.com/Mellanox/nic-configuration-operator/pkg/host"
+	"github.com/Mellanox/nic-configuration-operator/pkg/maintenance"
+	"github.com/Mellanox/nic-configuration-operator/pkg/types"
+)
+
+const nicDeviceSyncEventName = "nic-device-sync-event-name"
+
+var requeueTime = 1 * time.Minute
+
+// NicDeviceReconciler reconciles a NicDevice object
+type NicDeviceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	nodeName      string
+	namespaceName string
+
+	hostManager        host.HostManager
+	maintenanceManager maintenance.MaintenanceManager
+}
+
+type nicDeviceConfigurationStatuses []*nicDeviceConfigurationStatus
+
+type nicDeviceConfigurationStatus struct {
+	device                 *v1alpha1.NicDevice
+	nvConfigUpdateRequired bool
+	rebootRequired         bool
+	lastStageError         error
+}
+
+// Reconcile reconciles the NicConfigurationTemplate object
+func (r *NicDeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
I think it makes sense to put the reconciliation in multiple helper methods. Otherwise, it can pretty hard to read and maintain it. WDYT?

Reply:
Broken down the reconcile func a bit, hope you like it now :)